ClearKeep/Scenes/Home/Models/CallViewModel.swift

```diff
@@ -0,0 +1,454 @@
+			callTimer.fire()
+			
+			// => Ko check timeout nữa
+			self.stopTimeoutTimer()
+		}
+	}
+	
+	@objc private func timeTick(timer: Timer) {
+		let timeNow = timeCounter.timeNow()
+		timeCall = timeNow
+	}
+	
+	private func stopCallTimer() {
+		CFRunLoopStop(CFRunLoopGetCurrent())
+		callTimer?.invalidate()
+		callTimer = nil
+	}
+	
+	private func stopTimeoutTimer() {
+		CFRunLoopStop(CFRunLoopGetCurrent())
+		timeoutTimer?.invalidate()
+		timeoutTimer = nil
+	}
+	
+	@objc private func checkCallTimeout() {
+		print("checkCallTimeout \(callInterval)")
+		callInterval += 1
+		if callInterval >= 60 && callTimer == nil {
+			endCall()
+		}
+	}
+	
+	func updateCallTypeVideo() {
+		guard let callBox = self.callBox else { return }
+		DispatchQueue.main.async { [weak self] in
+			guard let self = self else { return }
+			self.cameraOn = true
+			self.callType = .video
+			callBox.isCamera = true
+			callBox.type = .video
+			callBox.videoRoom?.publisher?.cameraOn()
+			self.speakerEnable = true
+			self.updateSpeakerConfig()
+			print("#TEST updateCallTypeVideo >>> video type")
+		}
+	}
+	
+	func updateCallType(data: PublicationNotification) {
+		guard let callBox = self.callBox,
+			  callBox.roomRtcId == Int64(data.groupID ?? "0")
+		else {
+			return
+		}
+		
+		DispatchQueue.main.async { [weak self] in
+			guard let self = self else { return }
+			self.cameraOn = false
+			self.callType = data.notifyType == "video" ? .video : .audio
+			self.callBox?.type = self.callType == .video ? .video : .audio
+			callBox.isCamera = false
+			if self.callType == .video {
+				self.callBox?.videoRoom?.publisher?.cameraOff()
+			}
+			self.speakerEnable = true
+			self.updateSpeakerConfig()
+			print("#TEST updateCallTypeVideo >>> video type")
+		}
+	}
+	
+	func didReceiveMessageGroup(userInfo: [AnyHashable : Any]?) {
+		print("#TEST didReceiveMessageGroup >>>> userInfo: \(userInfo?.debugDescription ?? "")")
+//		if let userInfo = userInfo,
+//		   let publication = userInfo["publication"] as? Notification_NotifyObjectResponse {
+//			if publication.notifyType == "audio" ||  publication.notifyType == "video" {
+//				if publication.notifyType == "video" {
+//					DispatchQueue.main.async { [weak self] in
+//						guard let self = self else { return }
+//						self.callType = .video
+//						self.callBox?.type = .video
+//						self.speakerEnable = true
+//						self.updateSpeakerConfig()
+//						//self.callBox?.videoRoom?.publisher?.cameraOn()
+//						print("#TEST didReceiveMessageGroup >>>> video type")
+//					}
+//				}
+//				// TODO: Check audio type update
+//			}
+//		}
+	}
+}
+
+extension CallViewModel {
+	
+	func videoViewConfig(for videoView: RTCMTLEAGLVideoView) -> CustomVideoViewConfig {
+		for (key, value) in remotesVideoViewDict where value == videoView {
+			if let config = remotesVideoViewConfig[key] {
+				return config
+			}
+		}
+		
+		return CustomVideoViewConfig(clientId: "", groupId: 0)
+	}
+}
+
+
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ClearKeep/ck-ios/blob/fe9f474e9dacdb67166a3b18715e917932515a0f/ClearKeep/Scenes/Home/Models/CallViewModel.swift#L412"></span>Limit vertical whitespace to a single empty line. Currently 2.<br />
<code>vertical_whitespace</code> <code>CallViewModel.swift:412</code></td>
    </tr>
  </tbody>
</table>